replicatedhq/ship
# Ship init supports arbitrary upstreams

What I Did
------------

Updated how we resolve releases so that in addition to helm charts,
ship can pull an arbitrary directory of k8s manifests.

We now clone the repo into .ship, and then inspect if there's
a `Chart.yaml`. If there is, we moved the cloned repo path into `./chart`,
and generate the values-kustomize lifecycle. If there's no `Chart.yaml`,
we move it straight to `base`, and use a minimal lifecycle to drop the user
straight into "kustomize".

This is mostly a POC, and could probably use some tweaks on the
UX side, esp around UI messages and their pacing/verbosity.

It could probably also use some unit testing love, but this has good
integration test coverage for now.

Resolves #268 

How I Did it
------------

We've add a new workflow for resolving releases, to try and consolidate
some of the various levels of abstraction in the `specs.Resolver` and in
`kustomize.go`. In order:

- fetch upstream from github or using `helm fetch`
- determine application type based on the upstream we fetched. For now, type is one of `replicated.app`, `helm` or `k8s`, but i think we wanna add knative soon
- Write out upstream and content sha to state and to ShipAppMetadata
- If helm, add helm chart metadata to ShipAppMetadata as well
- If helm+fetch, write helm fetch parameters like `--chart-version` to state.json
- generate a release based on application type + discovered metadata

There is a *lot* of stuff moving around in this PR with respect to "how
do we get the release we hand off to daemon/lifecycle", it may be easier to
read the just the new code first, before diving into the diff.

Changes in order of appearance:

- refactor replicated.app spec fetching into specs/replicatedapp
- refactor github client into specs/githubclient
- new specs/apptype package for determining app type
- move `init_chart` integration tests to `init`, which just allows for passing in arbitrary args like `--chart-version` etc.
- rename `chartURL` to upstream everywhere
- rename `HelmChartMetadata` to `ShipAppMetadata`, the this info is
  pretty generic and can apply to non-helm charts
- Create a `.ship/tmp` for temporary workspace stuff, to be cleaned up
when ship terminates. This lets components write to that dir, without
having to worry about how/when to do a deferred cleanup.
  - Cobra commands do an os.MkdirAll on `.ship/tmp` before ship starts
  - Cobra commands do an os.RemoveAll on `.ship/tmp` as ship shuts down
  - On Shutdown, ship does an `os.RemoveAll` on `.ship/tmp` (this handles the case where cobra doesn't get to clean up, because ship did an os.Exit)
- Rename some constants to be more generic for non-helm apps
- use `BackupIfPresent` before removing any directories, move the
function into a utililty package, add a terminal warning when backing up
directories.
- Rework specs.Resolver for the new workflow above
- expand determineApplicationType to handle raw k8s case, as well as the old `helm fetch` case
- add methods to migrate `state.ChartURL` to `state.Upstream` cleanly.
- `helm` asset uses `.ship/tmp/.helm` as HELM_HOME

Other fun stuff:

- integration tests pretty-print json before diffing
- `state.Manager` saves pretty-printed state json
- Increase timeout from 60ms to 120ms on a flaky test
- More fixes for weird tmpdir stuff on osx

How to verify it
------------

Check out integration tests, but you can now do `ship init` for repo paths that
have no helm chart, just k8s yaml. For example, you can try out [my fork of
sysdig cloud's kubernetes installer](https://github.com/dexhorthy/sysdigcloud-kubernetes/tree/master/sysdigcloud):

```
ship init github.com/dexhorthy/sysdigcloud-kubernetes/sysdigcloud
```

Description for the Changelog
------------

Support fetching/kustomizing an arbitary directory of Kubernetes manifests with `ship init`.

![](https://upload.wikimedia.org/wikipedia/commons/thumb/a/a4/FrotamericaShipWreck.jpg/2560px-FrotamericaShipWreck.jpg)

Special thanks to @jadengore who helped out on debugging and fixing a few things here,
and whose commits got obliterated in what turned out to be a pretty messy rebase after
the changes for `helm fetch`.
